worlds/marioland2/logic.py

```diff
@@ -0,0 +1,608 @@
+    return has_pipe_right(state, player)
+
+
+def mario_zone_4_coins(state, player, coins):
+    return coins <= 63 or not is_auto_scroll(state, player, "Mario Zone 4")
+
+
+def not_blocked_by_sharks(state, player):
+    sharks = [state.multiworld.worlds[player].sprite_data["Turtle Zone 1"][i]["sprite"]
+              for i in (27, 28)].count("Shark")
+    if state.has("Carrot", player) or not sharks:
+        return True
+    if sharks == 2:
+        return state.has_all(["Mushroom", "Fire Flower"], player)
+    if sharks == 1:
+        return state.has_any(["Mushroom", "Fire Flower"], player)
+    return False
+
+
+def turtle_zone_1_normal_exit(state, player):
+    return not_blocked_by_sharks(state, player)
+
+
+def turtle_zone_1_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Turtle Zone 1")
+    reachable_coins = 30
+    if not_blocked_by_sharks(state, player):
+        reachable_coins += 13
+        if auto_scroll:
+            reachable_coins -= 1
+    if state.has("Water Physics", player) or state.has("Carrot", player):
+        reachable_coins += 10
+    if state.has("Carrot", player):
+        reachable_coins += 24
+        if auto_scroll:
+            reachable_coins -= 10
+    return coins <= reachable_coins
+
+
+def turtle_zone_2_normal_exit(state, player):
+    return (has_pipe_up(state, player) and has_pipe_down(state, player) and has_pipe_right(state, player) and
+            has_pipe_left(state, player) and state.has("Water Physics", player)
+            and not is_auto_scroll(state, player, "Turtle Zone 2"))
+
+
+def turtle_zone_2_secret_exit(state, player):
+    return (has_pipe_up(state, player) and state.has("Water Physics", player)
+            and not is_auto_scroll(state, player, "Turtle Zone 2"))
+
+
+def turtle_zone_2_midway_bell(state, player):
+    return turtle_zone_2_secret_exit(state, player) or state.has("Turtle Zone 2 Midway Bell", player)
+    
+    
+def turtle_zone_2_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Turtle Zone 2")
+    reachable_coins = 2
+    if auto_scroll:
+        if state.has("Water Physics", player):
+            reachable_coins += 6
+    else:
+        reachable_coins += 2
+        if state.has("Water Physics", player):
+            reachable_coins += 20
+        elif state.has("Turtle Zone 2 Midway Bell", player):
+            reachable_coins += 4
+        if (has_pipe_right(state, player) and has_pipe_down(state, player)
+                and state.has_any(["Water Physics", "Turtle Zone 2 Midway Bell"], player)):
+            reachable_coins += 1
+            if has_pipe_left(state, player) and has_pipe_up(state, player):
+                reachable_coins += 1
+                if state.has("Water Physics", player):
+                    reachable_coins += 1
+    return coins <= reachable_coins
+
+
+def turtle_zone_secret_course_normal_exit(state, player):
+    return state.has_any(["Fire Flower", "Carrot"], player)
```

Comment:
I could potentially be wrong but doesn't this strictly require Fire Flower, since Carrot has no way of breaking blocks?

Reply:
You don't need to break the blocks if you're small, and then you can reach the carrot.